lib/spree/graphql/lazy_resolvers/batch_loader/variant.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+class Spree::GraphQL::LazyResolvers::BatchLoader::Variant < Spree::GraphQL::LazyResolvers::Base
+  def images(query:)
+    query = Spree::GraphQL::Schema::Inputs::RansackQuery.queries_to_ransack_query(query)
+
+    ::BatchLoader::GraphQL.for(object.id).batch(default_value: []) do |variant_ids, loader|
+      images_batch(variant_ids, query).each do |image|
+        loader.call(image.viewable_id) { |memo| memo << image }
+      end
+    end
+  end
+
+  def option_values
+    ::BatchLoader::GraphQL.for(object.id).batch(default_value: []) do |variant_ids, loader|
+      option_values_batch(variant_ids).each do |option_values_variant|
+        loader.call(option_values_variant.variant_id) { |memo| memo << option_values_variant.option_value }
+      end
+    end
+  end
+
+  def price
+    pricing_options = context[:helpers].current_pricing_options
+
+    ::BatchLoader::GraphQL.for(object.id).batch(default_value: nil) do |variant_ids, loader|
+      price_batch(variant_ids).each do |price|
+        loader.call(price.variant_id) do |memo|
+          return memo if memo
+
+          price.try! :money if current_price?(price, pricing_options)
+        end
+      end
+    end
+  end
+
+  def prices
+    ::BatchLoader::GraphQL.for(object.id).batch(default_value: []) do |variant_ids, loader|
+      prices_batch(variant_ids).each do |price|
+        loader.call(price.variant_id) { |memo| memo << price.money }
+      end
+    end
+  end
+
+  def product
+    ::BatchLoader::GraphQL.for(object.product_id).batch do |product_ids, loader|
+      product_batch(product_ids).each { |product| loader.call(product.id, product) }
+    end
+  end
+
+  private
+
+  def images_batch(variant_ids, query)
+    Spree::Image.where(viewable_type: object.class.to_s, viewable_id: variant_ids).ransack(query).result
+  end
+
+  def option_values_batch(variant_ids)
+    Spree::OptionValuesVariant
+      .includes(:variant, :option_value)
+      .references(:option_value)
+      .where(variant_id: variant_ids)
+      .order(Spree::OptionValue.arel_table[:position].asc)
+  end
+
+  def price_batch(variant_ids)
```

Comment:
Why this method is needed? I think it's better to remove it and call `prices_batch` directly since this makes me think that will return a single price but it's returning an "array".

Reply:
The reason why I left that method is that I created all the `_batch`methods more for decorating purposes than for refactoring purposes; I'd like that if a developer wants to change the data retrieving (which is the most probable use case for reopening lazy resolvers classes to me), such developer is not forced to monkey patch all the logic but just the data retrieving methods. Now, if I call `prices_batch` directly in the field method instead of having a separated method for the `price` field, the developer who wishes to have a different implementation between the `prices` field batch and the `price` field batch will not be able to do it, and will be forced to monkeypatch the whole `price_batch` method.

But, as you noticed, the `*_batch` methods naming is a bit confusing because I'm substantially implying there `*_field_batch`: I mean that `price` doesn't stand for the returning value, but for the related field; maybe we should rename the `_batch` methods to `_field_batch`? WDYT?

Reply:
No, now it's clear. It's probably just me and my limited understanding of the GraphQL domain. I think it works!